### ISSUE TITLE
mousemove events lost when mouse leaves window during drag operation

### ISSUE THREAD

**Issue by mhalle** [Kitware/vtk-js#1671]
In all the vtkjs examples, drag operations such as panning or rotating (i.e., sequences of mousedown -> mousemove* -> mouseup events) end the drag operation when the mouse leaves the vtkjs canvas element. This limitation prevents interactions in smaller canvases that benefit from larger mouse or pointer motions. 

All contemporary browsers continue to pass mousemove events that happen during this kind of drag to the canvas, even while the mouse is outside the window.  vtkjs must be eating the event somewhere.

**Comment by jourdain** [comment]
Just provide a larger container when you call [this](https://github.com/Kitware/vtk-js/blob/master/Examples/Geometry/SimpleCone/index.js#L68)

**Comment by mhalle** [comment]
A larger container is only a partial fix.  Web browsers allow the event to propagate even outside the browser window. If a canvas is on the edge of the browser window, the event will get lost when the mouse leaves the browser window. 

So much expected web interaction depends on this property. For example scrolling or text selection would be much harder to use if mousemotion events during a drag ended at the canvas or window boundary.

**Comment by jourdain** [comment]
Are you saying that if your provide the `document` that does not solve your issue?

**Comment by mhalle** [comment]
I don't think it does. The problem happens with all of the vtkjs examples that use full screen render. Try and spin an object by clicking and holding the mouse and leave the window. The drag stops.  Try this same thing for this random demo:
https://playground.babylonjs.com/#1OH09K#131
The interaction doesn't stop when the mouse leaves the window. This behavior is standard and how events work by default in essentially all web browsers. There's something about vtkjs that's losing those events. 

This behavior has been true in most browsers forever. See this old stack overflow article about getting the same behavior out of an ancient version of IE, which was one of the only holdouts: https://stackoverflow.com/questions/1685326/responding-to-the-onmousemove-event-outside-of-the-browser-window-in-ie

**Comment by jourdain** [comment]
The other part is also related to the fact that if you really bound the event to a specific container, you could loose the outside events. And to handle that case we have this [listener](https://github.com/Kitware/vtk-js/blob/master/Sources/Rendering/Core/RenderWindowInteractor/index.js#L212) that automate the mouse up.

But this could be updated to accommodate the needs.

**Comment by jourdain** [comment]
The solution seems to just be

```
diff --git a/Sources/Rendering/Core/RenderWindowInteractor/index.js b/Sources/Rendering/Core/RenderWindowInteractor/index.js
index 3a29c58c37..a142f62da7 100644
--- a/Sources/Rendering/Core/RenderWindowInteractor/index.js
+++ b/Sources/Rendering/Core/RenderWindowInteractor/index.js
@@ -192,7 +192,7 @@ function vtkRenderWindowInteractor(publicAPI, model) {
   }
 
   function interactionRegistration(addListeners, force = false) {
-    const rootElm = document.querySelector('body');
+    const rootElm = document;
     const method = addListeners ? 'addEventListener' : 'removeEventListener';
     const invMethod = addListeners ? 'removeEventListener' : 'addEventListener';
 
@@ -209,7 +209,7 @@ function vtkRenderWindowInteractor(publicAPI, model) {
       }
 
       rootElm[method]('mouseup', publicAPI.handleMouseUp);
-      rootElm[method]('mouseleave', publicAPI.handleMouseUp);
+      // rootElm[method]('mouseleave', publicAPI.handleMouseUp);
       rootElm[method]('mousemove', publicAPI.handleMouseMove);
       rootElm[method]('touchend', publicAPI.handleTouchEnd, false);
       rootElm[method]('touchcancel', publicAPI.handleTouchEnd, false);
```

**Comment by jourdain** [comment]
https://github.com/Kitware/vtk-js/pull/1672

**Comment by mhalle** [comment]
Actually, I think the issue is solvable with judicious use of "setPointerCapture" during drag events. That seems to be what babylonjs uses in its event handlers. That guarantees that events continued to be captured during event sequences. The down event directs all events to the appropriate target, the up event (which is then guaranteed to be received) cancels the capture.
https://developer.mozilla.org/en-US/docs/Web/API/Element/setPointerCapture

**Comment by floryst** [comment]
Indeed, the pointer capture functionality was added to allow for effectively confining a pointer inside a particular element. The associated PR linked above should be mostly the same, as it turns out `mouseup` can be triggered outside the window. (Not sure if this was originally spec-defined, or is a browser quirk.)

**Comment by jourdain** [comment]
This should be fixed with 15.5.1
